### PR TITLE
Avoid reflection when rethrowing exceptions

### PR DESCRIPTION
**Pull request by pkoenig10**
`Throwables.rewrap` uses reflection to try an reconstruct an exception of the same type with a different message.

### CODE REVIEW

File: atlasdb-client/src/main/java/com/palantir/atlasdb/transaction/impl/TransactionRetryStrategy.java [hunk 1]
```diff
@@ -86,10 +85,7 @@ public <V> void onRetry(Attempt<V> attempt) {
         try {
             return retryer.call(task::run);
         } catch (RetryException e) {
-            Throwable wrapped = Throwables.rewrap(
-                    String.format("Failing after %d tries.", e.getNumberOfFailedAttempts()),
-                    e.getLastFailedAttempt().getExceptionCause());
-            throw throwTaskException(wrapped);
```
Comment (pkoenig10) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/transaction/impl/TransactionRetryStrategy.java:92]: We could add a suppress exception here, but the retry exhaustion is already logged in `logAttempt` so it seemed unnecessary to mention it here.

File: leader-election-impl/src/main/java/com/palantir/paxos/SingleLeaderPinger.java [hunk 2]
```diff
@@ -287,21 +286,12 @@ private void throwIfInvalidSetup(
             return;
         }
 
-        IllegalStateException exception =
```
Comment (ergo14) [inline leader-election-impl/src/main/java/com/palantir/paxos/SingleLeaderPinger.java:290]: I am not familiar with the *.pref things referenced here. I am guessing these are outdated and relate to previous deployment models?

**Review by ergo14 — COMMENTED** [review]
Makes sense. It looks like  we use `Throwables::rewrap` to capture the current stack trace. Are we okay losing that? If not, we should probably capture the stacktrace in another way?

File: leader-election-impl/src/main/java/com/palantir/paxos/SingleLeaderPinger.java (hunk 2 above)
Comment (pkoenig10) [reply leader-election-impl/src/main/java/com/palantir/paxos/SingleLeaderPinger.java:290]: Yeah I'm also unfamiliar with this and am fairly certain this message is no longer relevant with our current deployment models.

**Comment by pkoenig10** [comment]
> Are we okay losing that? If not, we should probably capture the stacktrace in another way?

 See my comment above about adding a suppressed exception:

https://github.com/palantir/atlasdb/pull/6931#discussion_r1465295795

I think it's fine losing the stacktrace in these places. The only place we _might_ want it is the one in `PaxosStateLogImpl`, but we emit an explicit log line there, which will capture the location.

**Comment by ergo14** [comment]
> > Are we okay losing that? If not, we should probably capture the stacktrace in another way?
> 
> See my comment above about adding a suppressed exception:
> 
> [#6931 (comment)](https://github.com/palantir/atlasdb/pull/6931#discussion_r1465295795)
> 
> I think it's fine losing the stacktrace in these places. The only place we _might_ want it is the one in `PaxosStateLogImpl`, but we emit an explicit log line there, which will capture the location.

Gotcha, apologies, I misunderstood the comment.

**Review by ergo14 — APPROVED** [review]
👍🏼

**Pull request merged**